python/tvm/relay/ir_builder.py

```diff
@@ -0,0 +1,299 @@
+
+        def _on_exit():
+            bindings, _, _, ret_value = self.exit_scope()
+            body = _mk_let(bindings, ret_value)
+            pfunc.body = body
+
+        return WithScope(pfunc, _on_exit)
+
+    def ret(self, x):
+        if not self.ret_values[-1]:
+            self.ret_values[-1] = into_ast(x)
+        else:
+            raise Exception(
+                "return value already set, a function can only have one return value")
+
+    def if_scope(self, cond):
+        """Construct the if branch an if expression with scoping."""
+        self.enter_scope()
+
+        def _on_exit():
+            bindings, _, _, ret_value = self.exit_scope()
+            assert self.ret_values[-1] is None
+            true_branch = _mk_let(bindings, ret_value)
+            self.ret_values[-1] = If(cond, true_branch, None)
+
+        return WithScope(10, _on_exit)
+
+    def else_scope(self):
+        """Construct the else branch of an if expression with scoping."""
+        self.enter_scope()
+
+        def _on_exit():
+            bindings, _, _, ret_value = self.exit_scope()
+            partial_if = self.ret_values[-1]
+            assert isinstance(
+                partial_if, If) and partial_if.false_branch is None
+            false_branch = _mk_let(bindings, ret_value)
+            self.ret_values[-1] = If(
+                partial_if.cond,
+                partial_if.true_branch,
+                false_branch)
+
+        return WithScope(10, _on_exit)
+
+    def param(self, name, ty=None):
+        if not ty:
+            ty = float_type()
+
+        return Param(Var(name), ty)
+
+    # def params(*args):
+    #      i = 0
+    #      while i < args.size():
+    #          arg = args[i]
+    #          if isinstance(arg, str):
+
+    def global_var(self, name: str):
+        return self.env.global_var(name)
+
+    def decl(self, name: str, *params, ret_type=None):
+        self.enter_scope()
+
+        def _on_exit():
+            bindings, _, _, ret_value = self.exit_scope()
+            exp = _mk_let(bindings, ret_value)
+            self.env.add(name, Function(params, ret_type, exp))
+
+        return WithScope(10, _on_exit)
+
+    # def while_loop(cond)
+
+    def get(self):
+        """Get the full program"""
+        bindings = self.bindings.pop()
+        scope = self.scopes.pop()
+
+        if self.bindings:
+            raise Exception("IRBuilder: binding error")
+
+        if self.scopes:
+            raise Exception("IRBuilder: scoping error")
+
+        if bindings and scope and not self.ret_values:
+            raise Exception("IRBuilder: no return value set")
+
+        return _mk_let(bindings, self.ret_values[-1]), self.env
+
+
+def bool_dtype():
+    return 'uint1'
+
+
+def int_dtype(bits=32):
+    return f'int{bits}'
+
+
+def float_dtype(bits=32):
+    return f'float{bits}'
+
+
+def uint_dtype(bits=32):
+    return f'uint{bits}'
+
+
+def int_type(bits=32, _lanes=1):
```

Comment:
for python users, it is more convenient for them to directly construct TensorType((3,4), "float32x1") and ScalarType("float32x1"), maybe we can do a bit smart convention of scalar type for them